ecosystem/nft-metadata-crawler/src/asset_uploader/worker/mod.rs

```diff
@@ -0,0 +1,128 @@
+use serde::Deserialize;
+use std::{sync::Arc, time::Duration};
+use tracing::{error, info};
+use url::Url;
+
+pub mod config;
+
+#[derive(Clone)]
+pub struct AssetUploaderWorkerContext {
+    config: Arc<AssetUploaderWorkerConfig>,
+}
+
+#[derive(Debug, Deserialize)]
+struct UploadRequest {
+    url: Url,
+}
+
+impl AssetUploaderWorkerContext {
+    pub fn new(config: AssetUploaderWorkerConfig) -> Self {
+        Self {
+            config: Arc::new(config),
+        }
+    }
+
+    /// Uploads an asset to Cloudflare and returns the response
+    async fn upload_asset(&self, url: &Url) -> anyhow::Result<impl IntoResponse> {
+        let hashed_url = sha256::digest(url.to_string());
+        let client = Client::builder()
+            .timeout(Duration::from_secs(MAX_ASSET_UPLOAD_RETRY_SECONDS))
+            .build()
+            .context("Error building reqwest client")?;
+        let form = Form::new()
+            .text("id", hashed_url.clone()) // Replace with actual metadata
+            .text("url", url.to_string());
+
+        info!(
+            asset_uri = ?url,
+            "[Asset Uploader] Uploading asset to Cloudflare"
+        );
+
+        let res = client
+            .post(format!(
+                "https://api.cloudflare.com/client/v4/accounts/{}/images/v1",
+                self.config.cloudflare_account_id
+            ))
+            .header(
+                "Authorization",
+                format!("Bearer {}", self.config.cloudflare_auth_key),
+            )
+            .multipart(form)
+            .send()
+            .await
+            .context("Error sending request to Cloudflare")?;
+
+        reqwest_response_to_axum_response(res).await
+    }
+
+    async fn handle_upload(
+        Extension(context): Extension<Arc<AssetUploaderWorkerContext>>,
+        Json(request): Json<UploadRequest>,
+    ) -> impl IntoResponse {
+        match context.upload_asset(&request.url).await {
+            Ok(res) => {
+                let res = res.into_response(); // TODO: How to log response body?
+                info!(asset_uri = ?request.url, response = ?res, "[Asset Uploader] Asset uploaded with response");
+                res
+            },
+            Err(e) => {
+                error!(asset_uri = ?request.url, error = ?e, "[Asset Uploader] Error uploading asset");
+                (
+                    StatusCode::INTERNAL_SERVER_ERROR,
+                    format!("Error uploading asset: {}", e),
+                )
+                    .into_response()
+            },
+        }
+    }
+}
+
+impl Server for AssetUploaderWorkerContext {
+    fn build_router(&self) -> Router {
+        let self_arc = Arc::new(self.clone());
+        Router::new()
+            .route("/", post(Self::handle_upload))
+            .layer(Extension(self_arc.clone()))
+    }
+}
+
+/// Converts a reqwest response to an axum response
+/// Only copies the response status, response body, and Content-Type header
```

Comment:
wondering is it worth having these two libraries for incoming and outgoing http requests? what if we just a single rust library for requests? chatgpt suggested hyper?

Reply:
we previously used reqwest in nft netadata crawler, which is why i used reqwest here. Also hyper makes it very hard to submit multipart requests bc hyper too low level and doesnt support it out of the box